### PR TITLE
Sightings security

### PR DESCRIPTION
**Pull request by colinwkingen**
First pull from branch. Adds user to Encounter, Sightings model and connects Sighting to Encounters.

### CODE REVIEW

**Comment by colinwkingen** [comment]
Retracting to do better testing

File: app/modules/encounters/models.py [hunk 1]
```diff
@@ -5,7 +5,7 @@
 """
 
 from app.extensions import db, FeatherModel
-
+from app.modules.sightings.models import Sighting
```
Comment (hwindsor) [inline app/modules/encounters/models.py:8]: Importing models from other models at a module scope caused circular dependencies. 
JP changed them all to be function level imports on the rest of the code.

File: app/modules/sightings/models.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    guid = db.Column(
+        db.GUID, default=uuid.uuid4, primary_key=True
+    )  # pylint: disable=invalid-name
+    title = db.Column(db.String(length=50), nullable=False)
+
+    def __repr__(self):
+        return (
+            '<{class_name}('
+            'guid={self.id}, '
```
Comment (hwindsor) [inline app/modules/sightings/models.py:24]: boilerplate gets this wrong, that should be self.guid

File: app/modules/sightings/models.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    def get_owners(self):
+        owners = []
+        for encounter in self.get_encounters():
+            if encounter.get_owner() is not None and encounter.get_owner() not in owners:
+                owners.append(encounter.get_owner())
+        return owners
+
+    def get_owner(self):
+        #this is what we talked about but it makes me squeamish
```
Comment (hwindsor) [inline app/modules/sightings/models.py:43]: Agreed, I would certainly suggest a len != 0 check first

**Comment by hwindsor** [comment]
Looks good.
I would also suggest the addition of tasks for the sighting, firstly just to list them all,
See:
https://github.com/WildbookOrg/houston/blob/develop/tasks/app/encounters.py
I find these extremely useful to know what on earth is going on :)

File: app/modules/encounters/models.py (hunk 1 above)
Comment (colinwkingen) [reply app/modules/encounters/models.py]: Cool, I can see that pattern now. Resolved in this case and will remember for later.

File: app/modules/sightings/models.py (hunk 2 above)
Comment (colinwkingen) [reply app/modules/sightings/models.py]: Resolved.

File: app/modules/sightings/models.py (hunk 3 above)
Comment (colinwkingen) [reply app/modules/sightings/models.py:43]: I've now added that check, good idea. I meant more that I'm still a little uncomfortable with the simplicity... If we can have any number of owners, I'm not sure if it will ever be safe to just pick one off the top of the stack. Maybe additional logic to return the one with ownership of the most encounters on the sighting? Willing to discuss.

**Comment by colinwkingen** [comment]
Reopened with changes from @hwindsor 's feedback, and some added testing.

File: tests/modules/sightings/test_models.py [hunk 4]
```diff
@@ -1,48 +1,101 @@
...
+
+    db.session.refresh(test_encounter_a)
+    db.session.refresh(test_sighting)
+
+    assert len(test_sighting.get_encounters()) == 0
+    assert test_encounter_a.get_sighting() is None
+
+    test_sighting.add_encounter(test_encounter_a)
+
+    with db.session.begin():
```
Comment (hwindsor) [inline tests/modules/sightings/test_models.py:35]: test_encounter_a and test_sighting were added to the db above, they shouldn't need readiing once linked together.... I don't think.

File: tests/modules/sightings/test_models.py [hunk 5]
```diff
@@ -1,48 +1,101 @@
...
+
+    test_encounter = Encounter(
+        title='New Test Encounter',
+    )
+
+    with db.session.begin():
+        db.session.add(test_encounter)
+        db.session.add(test_sighting)
+
+    db.session.refresh(test_encounter)
```
Comment (hwindsor) [inline tests/modules/sightings/test_models.py:72]: I might be wrong but I think you need the with db.session.begin() bit for refreshes as well.

File: app/modules/sightings/models.py (hunk 3 above)
Comment (hwindsor) [reply app/modules/sightings/models.py:43]: Ah yes, I see what you mean. 
I think your idea of returning the owner of the most encounters is a sensible one. Sightings are different in that they can potentially have multiple owners.
Actually could they? Does this cause security issues in terms of 
User A creates encounter, with 3 sightings.
User B adds 4 of their sightings to the encounter and hence becomes the owner.
User B deletes encounter as it is now the owner.
Do we want to prevent a different user from the owner adding a sighting. This needs a bit more thought. What do you allow in old world.

In terms of the implementation, get_owners is only called from this method so I'd remove it and move the encounters list iteration into here and figure out the owner here

**Review by hwindsor — DISMISSED** [review]
Looking good, a couple of points about the tests, not certain if you're right or me on this one.

The point about the sighting ownership is possibly best left for a subsequent change after there's been a bit more discussion about how sightings should be owned with @naknomum and @bluemellophone

File: tests/modules/encounters/test_models.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+# pylint: disable=invalid-name,missing-docstring
+import json, logging
+
+
+def test_encounter_add_owner(db):
+
+    from app.modules.users.models import User
+    from app.modules.encounters.models import Encounter
+
+    test_user = User(
```
Comment (hwindsor) [inline tests/modules/encounters/test_models.py:11]: Better to use the utils.generate_user_instance() function

Comment (colinwkingen) [reply tests/modules/encounters/test_models.py:11]: Great! Now I know about that, thanks.

File: tests/modules/sightings/test_models.py (hunk 4 above)
Comment (colinwkingen) [reply tests/modules/sightings/test_models.py:35]: Cool. Will get some clarification on db transaction details in our next meeting, this was kinda hacked in and left in because it worked.

File: app/modules/sightings/models.py (hunk 3 above)
Comment (naknomum) [reply app/modules/sightings/models.py:43]: using "owner" in the sense of multi-owners (by way of sub-objects: encoutners) is a new concept for nextgen.  as such, i am not sure how to proceed -- but have some ideas.

i can see that we probably(!) pretty clearly want the .owner property to be the (original) creator of the sighting.  but right now the only place i can imagine caring about owner-by-way-of-encounters has to do with **security**, no?  so maybe multi-ownership (in terms of encounter-depth ownership) should be left to be resolved in _security-related_ methods on sightings?

in other words, we can imagine, say, making a decision about "who can delete a sighting?"  would this be based perhaps just one logic around .owner?  or would we want to somehow traverse the encounter-owners?  maybe sighting _read_ privilege would be based on our typical algorithms (based on .owner) but grant additional access if you happen to be an encounter-owner?  does that make sense?

in this way, the `get_owners()` method might really only be of interest "internally" for security checks.

Comment (naknomum) [reply app/modules/sightings/models.py:43]: though now i see the `user.owns_object(sighting)` is a little tricky based on what i say in previous comment.  for that to be an outright boolean concept (outside of context of "owner in what way?") is a tough call.  i am totally fine with it proceeding as written: basically mimicking old-world idea that if you owned an encounter _in a sighting_, thats good enough.  (i just honestly dont remember if that was "good enough" _to do anything_ to said sighting -- or if it was used only for _reading_ ?   anyway, lets just keep it as that for now and deal with later if we need something more subtle.

**Review by naknomum — DISMISSED** [review]
left a few comments in a couple ongoing conversations, but looks good to me.  this definitely should get into develop before i do my **enc_passthru** pull request, so i can deal with where there are conflicts from this within that branch.

File: app/modules/sightings/models.py (hunk 3 above)
Comment (colinwkingen) [reply app/modules/sightings/models.py:43]: In "old wildbook" we don't let you delete a sighting at all... it just hangs out until every encounter it contains is either removed or deleted then it just goes away. I think there is a great argument for that since it respects every participating user.

File: app/modules/encounters/models.py [hunk 7]
```diff
@@ -17,13 +16,27 @@ class Encounter(db.Model, FeatherModel):
     guid = db.Column(
         db.GUID, default=uuid.uuid4, primary_key=True
     )  # pylint: disable=invalid-name
+
+    from app.modules.sightings.models import Sighting
```
Comment (bluemellophone) [inline app/modules/encounters/models.py:20]: Is this import needed here?

File: tests/modules/encounters/test_models.py (hunk 6 above)
Comment (bluemellophone) [reply tests/modules/encounters/test_models.py:10]: I agree with @hwindsor, we should be using utility functions like this as much as we can

**Review by bluemellophone — APPROVED** [review]
This PR LGTM but I would like to see the database updates merged into a single update before merging into develop.

File: app/modules/encounters/models.py (hunk 1 above)
Comment (bluemellophone) [reply app/modules/encounters/models.py]: Yes, the design decision is we want config.py to control and define the import order of modules.  This defines how database tables are created and the SQLAlchemy ORM structures are registered.  If we move the imports out of module level and into function and class level, we can preserve the correct order.

**Review by bluemellophone — APPROVED** [review]
LGTM
